engine/flink/management/periodic/src/main/scala/pl/touk/nussknacker/engine/management/periodic/Utils.scala

```diff
@@ -1,13 +1,72 @@
 package pl.touk.nussknacker.engine.management.periodic
 
+import akka.actor.{ActorNotFound, ActorPath, ActorRef, ActorSystem}
+import akka.pattern.{AskTimeoutException, gracefulStop}
 import com.typesafe.scalalogging.LazyLogging
 
+import java.util.concurrent.TimeoutException
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, Future}
+
 object Utils extends LazyLogging {
 
+  private val GracefulStopTimeout         = 5 seconds
+  private val ActorResolutionTimeout      = 5 seconds
+  private val TotalActorResolutionTimeout = 16 seconds
+
   def runSafely(action: => Unit): Unit = try {
     action
   } catch {
     case t: Throwable => logger.error("Error occurred, but skipping it", t)
   }
 
+  def gracefulStopActor(actorRef: ActorRef, actorSystem: ActorSystem): Unit = {
+    logger.info(s"Gracefully stopping $actorRef")
+
+    try {
+      Await.result(
+        gracefulStop(actorRef, GracefulStopTimeout),
+        GracefulStopTimeout + (1 second)
+      )
+    } catch {
+      case _: AskTimeoutException | _: TimeoutException =>
+        throw new IllegalStateException(s"Failed to gracefully stop actor $actorRef within timeout")
+    }
+
+    logger.info(s"Gracefully stopped $actorRef, waiting for it's name to be freed")
+
+    try {
+      Await.result(
+        waitUntilActorNameIsFree(actorRef.path, actorSystem),
+        TotalActorResolutionTimeout
+      )
+    } catch {
+      case _: TimeoutException =>
+        throw new IllegalStateException(s"Failed to free actor path ${actorRef.path} within timeout")
+    }
+
+    logger.info(s"$actorRef is stopped and it's name is free")
```

Comment:
This part with `it's name is free` will look enigmatic for someone who just looking into the logs with no idea what's going on in the code. I think that `$actorRef was gracefully stopped` is enough on the `info` level.